### PR TITLE
Bright edges source code and demo

### PR DESCRIPTION
**Pull request by fiammante**
A new approach for visualizing edges. It uses the absolute difference of 2 blurs to compute gradients creating local minimas for edges. The histogram equalization makes the edges appears darker within a bright zone.
If a contrast value of zero is given no additional process is performed. 
Otherwise a contrasting correction finds the local minimas and sets the edge to black, while setting the surrounding to white. A pixel correction restores line continuity and removes dark spots.
The default contrast value is 1 over a range of 255. This default contrast ensures visualizing edges on chessboard like images or low resolution images where small objects edges may be needed.
chessprocess
Other contrast values will look for deeper local minimas, removing smaller details and only leaving higher gradient variations.
testbrightedges
The images in this entry are the result of brightedgesexample.cpp .

### CODE REVIEW

**Comment by fiammante** [comment]
@tenderfail I applied all of your recommendations to this pull request new branch, edge.ptr() etc.  Please let me know if it is now complete.

File: modules/ximgproc/src/brightedges.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+		if (line < weight) lines += 1;
+		if (line == 1) return 0;
+		int surround = iedge.ptr(row - 1)[col - 1] +
+			iedge.ptr(row - 1)[col] +
+			iedge.ptr(row - 1)[col + 1] +
+			iedge.ptr(row)[col - 1] +
+			iedge.ptr(row)[col + 1] +
+			iedge.ptr(row + 1)[col - 1] +
+			iedge.ptr(row + 1)[col] +
+			iedge.ptr(row - 1)[col + 1];
```
Comment (terfendail) [inline modules/ximgproc/src/brightedges.cpp]: Shouldn't it be `iedge.ptr(row + 1)[col + 1]`?

**Comment by fiammante** [comment]
@terfendail Good catch! Just fixed it.

**Comment by fiammante** [comment]
@terfendail I don't understand why sometimes the build fails on test_stereo which I don't use. Can you help me?

File: modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+#include "opencv2/core.hpp"
+namespace cv
+{
+	CV_EXPORTS_W void  BrightEdges(Mat &_original, Mat &_edgeview, int contrast = 1, int shortrange = 3, int longrange = 9);
+	int correctPixel(Mat &iedge, int row, int col);
+	int contrastEdges(Mat &minput, Mat &mouput, int contrast);
+	bool isPixelMinimum(Mat &edge, int row, int col, int contrast);
```
Comment (terfendail) [inline modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp]: Last three functions look like internal.
Could you please remove their definition from the header?

File: modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+#include "opencv2/core.hpp"
+namespace cv
+{
+	CV_EXPORTS_W void  BrightEdges(Mat &_original, Mat &_edgeview, int contrast = 1, int shortrange = 3, int longrange = 9);
```
Comment (terfendail) [inline modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp]: Could you please move definition inside nested namespace `xfeatures2d`

**Comment by terfendail** [comment]
At least I could try to find out the reason of test_stereo failures but at the moment it isn't failed yet.

File: modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp (hunk 2 above)
Comment (fiammante) [reply modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp]: @terfendail doing it now. I have added them because build complained about missing prototype.

**Comment by terfendail** [comment]
Could you please add inclusion of `ximgproc/brightedges.hpp` to `ximgproc.hpp`

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+int main(int argc, const char** argv)
+{
+	CommandLineParser parser(argc, argv, keys);
+	if (parser.has("help"))
+	{
+		help();
+		return 0;
+	}
+	string filename = parser.get<string>(0);
+	image = imread(filename, 1);
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: It would be better to use `IMREAD_COLOR` constant here instead of exact value

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+	if (image.empty())
+	{
+		printf("Cannot read image file: %s\n", filename.c_str());
+		help();
+		return -1;
+	}
+	cedge.create(image.size(), image.type());
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: `cedge` matrix isn't used inside sample

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	cvtColor(image, gray, COLOR_BGR2GRAY);
+	// Create a window
+	// //  " original ";
+	namedWindow("Original");
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: It isn't necessary to precreate the window if there is no special requirement for window properties. `imshow` could create the window on the own.
IMHO it would be better to shorten the service part of the sample code to make cleaner the usage of described functionality

File: modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp (hunk 3 above)
Comment (fiammante) [reply modules/ximgproc/include/opencv2/ximgproc/brightedges.hpp]: done. When I move out the prototypes of functions to the cpp I get "warning: no previous declaration for"  xxx. But I have them on top of the file. I don't understand why I get this warning.

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 4 above)
Comment (fiammante) [reply modules/ximgproc/samples/brightedgesexample.cpp]: done

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 5 above)
Comment (fiammante) [reply modules/ximgproc/samples/brightedgesexample.cpp]: removed

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	namedWindow("Contrast 5");
+	imshow("Contrast 5", edge);
+	// " Contrast 10  \n";
+	BrightEdges(image, edge, 10);
+	namedWindow("Contrast 10");
+	imshow("Contrast 10", edge);
+	//  "wait key ";
+	waitKey(0);
+	// //  "Destroy ";
+	destroyAllWindows();
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: Windows destruction is also unnecessary since it will be performed automatically on application termination

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+#include "opencv2/core/utility.hpp"
+#include "opencv2/imgproc.hpp"
+#include "opencv2/imgcodecs.hpp"
+#include "opencv2/highgui.hpp"
+#include <stdio.h>
+#include <iostream>
+#include "opencv2/ximgproc/brightedges.hpp"
+using namespace cv;
+using namespace std;
+int edgeThresh = 1;
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: edgeThresh isn't used inside sample

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 7 above)
Comment (fiammante) [reply modules/ximgproc/samples/brightedgesexample.cpp]: removed

**Comment by alalek** [comment]
> why sometimes the build fails on test_stereo

Because test_stereo is buggy and has sporadic failures (at least it uses `rand()` call instead of reproducible `cv::RNG`).

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+#include "opencv2/imgcodecs.hpp"
+#include "opencv2/highgui.hpp"
+#include <stdio.h>
+#include <iostream>
+#include "opencv2/ximgproc.hpp"
+using namespace cv;
+using namespace ximgproc;
+using namespace std;
+int edgeThresh = 1;
+Mat image, gray, gblur, bblur, edge, cedge,diff;
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: IMHO it would be better to move matrix declarations closer to the place they are used.

**Comment by fiammante** [comment]
@terfendail moved Mat declaration before place they are used.

File: modules/ximgproc/src/brightedges.cpp [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+#include "precomp.hpp"
+#include <iostream>
+#include <signal.h>
+namespace cv
+{
+	namespace ximgproc {
+
+	int correctPixel(Mat &iedge, int row, int col);
+	int contrastEdges(Mat &minput, Mat &mouput, int contrast);
+	bool isPixelMinimum(Mat &edge, int row, int col, int contrast);
```
Comment (terfendail) [inline modules/ximgproc/src/brightedges.cpp]: IMHO it would be better to define functions as static instead of forward declarations since they aren't used outside of the module

Comment (fiammante) [reply modules/ximgproc/src/brightedges.cpp]: Removed unused Mat and moved declaration before where they are used.

**Comment by fiammante** [comment]
@terfendail changed local functions to static

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 8 above)
Comment (terfendail) [reply modules/ximgproc/samples/brightedgesexample.cpp]: Could you please remove `edgeThresh` variable since it is unused.
Also could you please replace tabs with spaces. OpenCV standard for indentation is [4 spaces](https://github.com/opencv/opencv/wiki/Coding_Style_Guide)

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 6 above)
Comment (terfendail) [reply modules/ximgproc/samples/brightedgesexample.cpp]: I think it would be better to remove`namedWindow` calls which are unrelated to the algorithm and just clutter the code

Comment (fiammante) [reply modules/ximgproc/samples/brightedgesexample.cpp]: @terfendail without namedwindow the windows just do not stay displayed on Windows 10. Just like described here http://answers.opencv.org/question/135303/namedwindow-imshow-not-showing-on-the-screen/  am I missing something?

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 8 above)
Comment (fiammante) [reply modules/ximgproc/samples/brightedgesexample.cpp]: @terfendail and removed unused variable

File: modules/ximgproc/samples/brightedgesexample.cpp (hunk 6 above)
Comment (terfendail) [reply modules/ximgproc/samples/brightedgesexample.cpp]: @fiammante mentioned question is related to an issue of specific setup rather than inappropriate way to create window an moreover it doesn't work properly even with `namedWindow`.
I've tried to reproduce the issue on Win10 and at least on my machine window appears normally both with `namedWindow` and without it. However I understand that doesn't mean it works everywhere.
Anyway image creation is unrelated to this PR so I think it could be merged.
BTW `imshow` should be able to create the window on the own so if it doesn't could you please create the issue describing it with details of your setup

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+#include "opencv2/imgproc.hpp"
+#include "opencv2/imgcodecs.hpp"
+#include "opencv2/highgui.hpp"
+#include <stdio.h>
+#include <iostream>
+#include "opencv2/ximgproc.hpp"
+using namespace cv;
+using namespace ximgproc;
+using namespace std;
+// define a trackbar callback
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: This comment looks unrelated

File: modules/ximgproc/samples/brightedgesexample.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    //  " absdiff ";
+    Mat edge;
+    BrightEdges(image, edge, 0); //  No contrast
+    namedWindow("Absolute Difference");
+    imshow("Absolute Difference", edge);
+    // " default contrast 1 ";
+    BrightEdges(image, edge);
+    namedWindow("Default contrast");
+    imshow("Default contrast", edge); // Default contrast 1
+    // Wait for a key stroke; the same function arranges events processing
```
Comment (terfendail) [inline modules/ximgproc/samples/brightedgesexample.cpp]: This comment looks unrelated

**Comment by fiammante** [comment]
@terfendail fixed 2 unrelated comments (leftover from previous code).

**Comment by terfendail** [comment]
👍  I think it could be merged now
@fiammante Thanks for contribution
